File: spring-data/src/main/java/dev/cerbos/queryplan/springdata/SizeTranslator.java
/*
 * Copyright 2021-2026 Zenauth Ltd.
 * SPDX-License-Identifier: Apache-2.0
 */

package dev.cerbos.queryplan.springdata;

import dev.cerbos.api.v1.engine.Engine.PlanResourcesFilter;
import dev.cerbos.api.v1.engine.Engine.PlanResourcesFilter.Expression.Operand;

import jakarta.persistence.criteria.CriteriaBuilder;
import jakarta.persistence.criteria.Expression;
import jakarta.persistence.criteria.Path;
import jakarta.persistence.criteria.Predicate;

import com.google.protobuf.Value;

import java.util.List;

/**
 * Translates {@code size(x) op N}: string length of a column, element count of a relation
 * chain, and the strict {@code size(filter(...))} count.
 *
 * <p>A direct relation may use the two-valued EXISTS shortcuts. A chain always uses a COUNT
 * guarded by {@link ChainSubqueries#requireLeadingHops}, so an absent to-one parent stays
 * UNKNOWN under both polarities.
 */
final class SizeTranslator {

    private final CriteriaBuilder cb;
    private final TriPredicate tri;
    private final PlanWalker walker;
    private final ChainSubqueries subqueries;

    SizeTranslator(CriteriaBuilder cb, TriPredicate tri, PlanWalker walker,
                   ChainSubqueries subqueries) {
        this.cb = cb;
        this.tri = tri;
        this.walker = walker;
        this.subqueries = subqueries;
    }

    /**
     * The constant, adjusted for an integral count. Truncating a fraction ({@code >= 1.5} as
     * {@code >= 1}) would return rows the PDP denies, so:
     * <ul>
     *   <li>{@code eq f} is always false and {@code ne f} always true, held in
     *       {@code decided};</li>
     *   <li>{@code gt f} and {@code ge f} become {@code ge ceil(f)};</li>
     *   <li>{@code lt f} and {@code le f} become {@code le floor(f)}.</li>
     * </ul>
     * An integral constant keeps its operator and {@code decided} is {@code null}.
     */
    private record Threshold(String op, long value, Boolean decided) {

        static Threshold of(String op, double raw) {
            if (raw == Math.rint(raw)) {
                return new Threshold(op, (long) raw, null);
            }
            return switch (op) {
                case "eq" -> new Threshold(op, (long) Math.floor(raw), Boolean.FALSE);
                case "ne" -> new Threshold(op, (long) Math.floor(raw), Boolean.TRUE);
                case "gt", "ge" -> new Threshold("ge", (long) Math.ceil(raw), null);
                case "lt", "le" -> new Threshold("le", (long) Math.floor(raw), null);
                // The planner only emits comparisons over size().
                default -> throw Refusals.malformed("Unsupported size comparison operator: " + op);
            };
        }
    }

    /**
     * The variable {@code size()} is taken of, plus the lambda for {@code size(filter(...))};
     * for a filter over a literal list, the list in place of the variable.
     */
    private record SizeArgument(String variable, ParsedLambda filter, Value literal) {

        static SizeArgument parse(PlanResourcesFilter.Expression sizeExpr) {
            List<Operand> sizeOps = sizeExpr.getOperandsList();
            if (sizeOps.size() != 1) {
                throw Refusals.malformed(
                        "Unsupported size() expression: size() takes exactly 1 argument, got "
                                + sizeOps.size());
            }
            Operand arg = sizeOps.get(0);
            if (arg.getNodeCase() == Operand.NodeCase.VARIABLE) {
                return new SizeArgument(arg.getVariable(), null, null);
            }
            String argOperator = arg.getNodeCase() == Operand.NodeCase.EXPRESSION
                    ? arg.getExpression().getOperator() : null;
            if ("filter".equals(argOperator)) {
                List<Operand> filterOps = arg.getExpression().getOperandsList();
                if (filterOps.size() != 2) {
                    throw Refusals.malformed("Unsupported size(filter(...)) expression");
                }
                Operand collection = filterOps.get(0);
                boolean literal = collection.getNodeCase() == Operand.NodeCase.VALUE;
                if (!literal && collection.getNodeCase() != Operand.NodeCase.VARIABLE) {
                    // filter() over a computed collection: legal CEL with no join chain.
                    throw Refusals.unsupported("Unsupported size(filter(...)) expression");
                }
                return new SizeArgument(literal ? null : collection.getVariable(),
                        ParsedLambda.parse(filterOps.get(1),
                                "Unsupported size(filter(...)) expression",
                                "lambda requires exactly 2 operands",
                                "lambda requires exactly 2 operands"),
                        literal ? collection.getValue() : null);
            }
            if ("except".equals(argOperator)) {
                // size(coll.except([...])): list difference has no JPA translation.
                throw Refusals.exceptUnsupported();
            }
            // A computed collection, e.g. a map() projection or a literal list.
            throw Refusals.unsupported(
                    "Unsupported size() expression: size() argument must be a collection "
                            + "attribute or filter(...), got " + Refusals.describeOperand(arg));
        }
    }

    /**
     * Translates {@code op(size(...), N)}, or returns {@code null} when there is no
     * {@code size()} operand or numeric constant. Operands must already be field-first.
     */
    Predicate trySizeComparison(String op, List<Operand> operands, Scope scope) {
        // Every leaf comparison passes through here, so avoid converting values until a
        // size() operand is found.
        PlanResourcesFilter.Expression sizeExpr = null;
        Double constant = null;
        for (Operand o : operands) {
            if (o.getNodeCase() == Operand.NodeCase.EXPRESSION
                    && "size".equals(o.getExpression().getOperator())) {
                sizeExpr = o.getExpression();
            } else if (o.getNodeCase() == Operand.NodeCase.VALUE
                    && o.getValue().getKindCase() == Value.KindCase.NUMBER_VALUE) {
                constant = o.getValue().getNumberValue();
            }
        }
        if (sizeExpr == null || constant == null) {
            return null;
        }

        Threshold threshold = Threshold.of(op, constant);
        SizeArgument arg = SizeArgument.parse(sizeExpr);
        if (arg.literal() != null) {
            return literalMatchCount(arg, threshold, scope);
        }
        Scope.Resolution resolved = scope.resolve(arg.variable());
        if (!(resolved instanceof Scope.ResolvedRelation ref)) {
            return stringLength(arg, (Scope.ResolvedScalar) resolved, threshold, scope);
        }
        if (arg.filter() == null) {
            return elementCount(ref, threshold, scope);
        }
        return matchingElementCount(ref, arg.filter(), threshold, scope);
    }

    /** {@code LENGTH(column) op N}. */
    private Predicate stringLength(SizeArgument arg, Scope.ResolvedScalar scalar,
                                   Threshold threshold, Scope scope) {
        // A bare lambda element also resolves as a scalar, but its mapping is a Relation and
        // has no string length.
        if (!(scalar.mapping() instanceof AttributeMapping.Field)) {
            throw Refusals.unmapped(
                    "size() requires a collection (Relation) mapping for " + arg.variable());
        }
        if (arg.filter() != null) {
            throw Refusals.unmapped("size(filter(...)) requires a collection (Relation) mapping for "
                    + arg.variable());
        }
        Path<?> path = scope.path(arg.variable());
        if (!String.class.equals(path.getJavaType())) {
            return tri.unknown();
        }
        // A decided arm must stay UNKNOWN for a NULL column (a missing attribute, which CEL
        // denies under both polarities): IS NOT NULL and a bare FALSE both flip under NOT.
        if (threshold.decided() != null) {
            return presentStringFold(path, threshold.decided());
        }
        // cb.length is an Integer expression, and narrowing an out-of-range threshold would
        // wrap (4294967296 becomes 0). No string length leaves int range, so those fold here.
        if (threshold.value() > Integer.MAX_VALUE) {
            return switch (threshold.op()) {
                case "eq", "gt", "ge" -> presentStringFold(path, false);
                case "lt", "le", "ne" -> presentStringFold(path, true);
                default -> throw Refusals.malformed(
                        "Unsupported size comparison operator: " + threshold.op());
            };
        }
        if (threshold.value() < Integer.MIN_VALUE) {
            return switch (threshold.op()) {
                case "gt", "ge", "ne" -> presentStringFold(path, true);
                case "eq", "lt", "le" -> presentStringFold(path, false);
                default -> throw Refusals.malformed(
                        "Unsupported size comparison operator: " + threshold.op());
            };
        }
        return compareCount(cb.length(path.as(String.class)), threshold.op(),
                (int) threshold.value());
    }

    /**
     * A {@code size(string)} comparison decided for every present string: {@code holds}, but
     * UNKNOWN when the column is NULL, so NOT of the fold still excludes the NULL rows. The
     * corpus's {@code size-huge-*-not} and {@code size-frac-*-not} actions pin both halves.
     */
    private Predicate presentStringFold(Path<?> path, boolean holds) {
        return tri.baseUnlessUnknown(holds ? cb.conjunction() : cb.disjunction(),
                () -> cb.isNull(path));
    }

    /**
     * {@code size(collection)}. No lambda is evaluated, so no element can be UNKNOWN.
     */
    private Predicate elementCount(Scope.ResolvedRelation ref, Threshold threshold, Scope scope) {
        if (threshold.decided() != null) {
            // Decided statically, but over a chain an absent to-one parent must stay UNKNOWN
            // under both polarities, so the answer is guarded by requireLeadingHops.
            if (!ref.isChained()) {
                return threshold.decided() ? cb.conjunction() : cb.disjunction();
            }
            return cb.equal(
                    subqueries.requireLeadingHops(scope, ref, cb.literal(1L), Long.class),
                    threshold.decided() ? 1L : 0L);
        }
        String op = threshold.op();
        long n = threshold.value();
        // The EXISTS shortcuts are two-valued: NOT EXISTS is TRUE for an absent parent. A chain
        // uses the guarded COUNT instead, which is NULL without the hop.
        if (!ref.isChained()) {
            boolean nonEmpty = ("gt".equals(op) && n == 0L) || ("ge".equals(op) && n == 1L);
            boolean empty = ("eq".equals(op) && n == 0L)
                    || ("le".equals(op) && n == 0L)
                    || ("lt".equals(op) && n == 1L);
            if (nonEmpty) {
                return anyElement(scope, ref);
            }
            if (empty) {
                return tri.not(anyElement(scope, ref));
            }
        }
        // A correlated COUNT, joined through every hop, so it counts the flattened tail
        // elements.
        return compareCount(
                subqueries.requireLeadingHops(scope, ref,
                        subqueries.countSubquery(scope, ref).sub(), Long.class),
                op, n);
    }

    private Predicate anyElement(Scope scope, Scope.ResolvedRelation ref) {
        return subqueries.existsSubquery(scope, ref, (sub, tailJoin, rebased) -> cb.conjunction());
    }

    /**
     * {@code size(coll.filter(x, pred))}. CEL filter errors if any element's predicate errors,
     * so the strict count is NULL when any body is UNKNOWN and the row is excluded under both
     * polarities.
     */
    private Predicate matchingElementCount(Scope.ResolvedRelation ref, ParsedLambda filter,
                                           Threshold threshold, Scope scope) {
        SubqueryBodyBuilder bodyBuilder = (sub, tailJoin, rebased) -> walker.traverse(
                filter.body(), Scope.lambda(tailJoin, sub, ref.tail(), filter.varName(), rebased));
        return walker.enterMacro("size(filter(...))", () -> {
            if (threshold.decided() != null) {
                // The count is decided, but an UNKNOWN body must still deny: the poison term is
                // 0 when every body is determined and NULL otherwise. It is also guarded for an
                // absent to-one parent.
                Expression<Long> poison = subqueries.requireLeadingHops(scope, ref,
                        subqueries.undeterminedPoisonSubquery(scope, ref, bodyBuilder),
                        Long.class);
                return threshold.decided() ? cb.equal(poison, 0L) : cb.notEqual(poison, 0L);
            }
            return compareCount(subqueries.strictMatchCountSubquery(scope, ref, bodyBuilder),
                    threshold.op(), threshold.value());
        });
    }

    /**
     * {@code size(filter(list, pred))} over a literal list, the planner's form for a list too
     * long to unroll: the strict count of {@link CollectionTranslator#literalStrictCount}, NULL
     * when any body is UNKNOWN. A decided threshold multiplies the count by zero, so it is
     * still UNKNOWN when poisoned.
     */
    private Predicate literalMatchCount(SizeArgument arg, Threshold threshold, Scope scope) {
        return walker.enterMacro("size(filter(...))", () -> {
            Expression<Long> count = CollectionTranslator.literalStrictCount(
                    cb, walker, arg.literal(), arg.filter(), scope);
            if (threshold.decided() != null) {
                Expression<Long> poison = cb.prod(count, 0L);
                return threshold.decided() ? cb.equal(poison, 0L) : cb.notEqual(poison, 0L);
            }
            return compareCount(count, threshold.op(), threshold.value());
        });
    }

    private <N extends Number & Comparable<N>> Predicate compareCount(
            Expression<N> count, String op, N n) {
        return switch (op) {
            case "eq" -> cb.equal(count, n);
            case "ne" -> cb.notEqual(count, n);
            case "lt" -> cb.lessThan(count, n);
            case "gt" -> cb.greaterThan(count, n);
            case "le" -> cb.lessThanOrEqualTo(count, n);
            case "ge" -> cb.greaterThanOrEqualTo(count, n);
            default -> throw Refusals.malformed(
                    "Unsupported size comparison operator: " + op);
        };
    }
}

File: exposed/src/main/kotlin/dev/cerbos/queryplan/exposed/SizeTranslator.kt
package dev.cerbos.queryplan.exposed

import com.google.protobuf.Value
import dev.cerbos.api.v1.engine.Engine.PlanResourcesFilter.Expression.Operand
import dev.cerbos.queryplan.exposed.sql.CountChars
import dev.cerbos.queryplan.exposed.sql.Params
import dev.cerbos.queryplan.exposed.sql.ScoreCase
import dev.cerbos.queryplan.exposed.sql.nullIfUndetermined
import org.jetbrains.exposed.v1.core.Alias
import org.jetbrains.exposed.v1.core.Coalesce
import org.jetbrains.exposed.v1.core.EqOp
import org.jetbrains.exposed.v1.core.Expression
import org.jetbrains.exposed.v1.core.GreaterEqOp
import org.jetbrains.exposed.v1.core.GreaterOp
import org.jetbrains.exposed.v1.core.IntegerColumnType
import org.jetbrains.exposed.v1.core.IsNullOp
import org.jetbrains.exposed.v1.core.LessEqOp
import org.jetbrains.exposed.v1.core.LessOp
import org.jetbrains.exposed.v1.core.Max
import org.jetbrains.exposed.v1.core.NeqOp
import org.jetbrains.exposed.v1.core.Op
import org.jetbrains.exposed.v1.core.PlusOp
import org.jetbrains.exposed.v1.core.Sum
import org.jetbrains.exposed.v1.core.Table
import org.jetbrains.exposed.v1.core.intLiteral
import kotlin.math.ceil
import kotlin.math.floor

/**
 * `size(x) op N`: string length over a column, element count over a relation chain, and the strict
 * `size(filter(...))` count. Owns the threshold arithmetic for fractional and out-of-range
 * constants, and the choice between the two-valued `EXISTS` emptiness shortcuts and the tri-state
 * count.
 *
 * It is a STEP of the comparison pipeline rather than a resolved-operand case: its SQL shapes are
 * subquery translations pinned by the differential oracle, not a (column, constant) pair.
 */
internal class SizeTranslator(private val translation: Translation) {

    /**
     * The comparison when one operand is a `size(...)` expression, or `null` when neither is.
     * Receives the NORMALISED operator and operands.
     */
    fun trySizeComparison(operator: String, operands: List<Operand>, scope: Scope): Op<Boolean>? {
        // The size() operand is detected before the constant is converted: every ordinary leaf
        // comparison probes through here, and materialising a list or a struct only to discard it
        // is work the common case should not pay for.
        val size = operands.lastOrNull {
            it.nodeCase == Operand.NodeCase.EXPRESSION && it.expression.operator == "size"
        }?.expression ?: return null
        val threshold = operands.lastOrNull {
            it.nodeCase == Operand.NodeCase.VALUE && it.value.kindCase == Value.KindCase.NUMBER_VALUE
        }?.value?.numberValue ?: return null

        val comparison = Threshold.of(operator, threshold)
        val argument = size.operandsList
        if (argument.size != 1) {
            throw Refusals.malformed("size() takes exactly 1 argument, got ${argument.size}")
        }
        return when (val target = SizeTarget.of(argument[0])) {
            is SizeTarget.Whole -> whole(target.variable, comparison, scope)
            is SizeTarget.Filtered -> filtered(target, comparison, scope)
            is SizeTarget.FilteredLiteral -> filteredLiteral(target, comparison, scope)
        }
    }

    /** `size(attribute)`: a string's length, or a relation's element count. */
    private fun whole(variable: String, comparison: Threshold, scope: Scope): Op<Boolean> =
        when (val resolved = scope.resolve(variable)) {
            is Resolution.Scalar -> stringLength(resolved, comparison)
            is Resolution.Collection -> elementCount(resolved, comparison)
        }

    /**
     * `size(string)`. The threshold arithmetic is decided statically rather than truncated: a
     * character count is integral, so a fractional constant can never be hit exactly, and
     * `LENGTH(s) >= 1` is a strictly wider filter than `size(s) >= 1.5` asks for.
     *
     * The vacuous arms are NOT unconditional. A NULL column is a missing attribute (or, under the
     * explicit convention, a null value with no `size()` overload), so CEL raises an evaluation
     * error and denies; folding to a constant TRUE would return those rows, and folding to a
     * constant FALSE would return them under a negation.
     */
    private fun stringLength(target: Resolution.Scalar, comparison: Threshold): Op<Boolean> {
        // `size()` counts characters, and CEL has no overload counting the characters of a number
        // or a boolean. Checked before the vacuous arms, which are answers about a string's length.
        if (translation.leaf.lacksTextOverload(target)) return TriLogic.unknown()
        translation.leaf.requireText("size()", target)
        val definite = { answer: Boolean ->
            TriLogic.baseUnlessUnknown(if (answer) Op.TRUE else Op.FALSE, IsNullOp(target.expression))
        }
        comparison.vacuous?.let { return definite(it) }
        // A string's length never leaves int range, so a threshold outside it decides the
        // comparison the same way a fractional one does — and an unguarded narrowing cast would
        // silently wrap it (2147483648 becomes -2147483648) and flip the filter.
        if (comparison.value > Int.MAX_VALUE) {
            return when (comparison.operator) {
                "eq", "gt", "ge" -> definite(false)
                else -> definite(true)
            }
        }
        if (comparison.value < Int.MIN_VALUE) {
            return when (comparison.operator) {
                "eq", "lt", "le" -> definite(false)
                else -> definite(true)
            }
        }
        return compare(CountChars(target.expression), comparison)
    }

    /**
     * `size(relation)`. Counting rows evaluates no lambda, so no element can be UNKNOWN and the
     * plain comparisons are already exact — except for what the CHAIN adds: an absent to-one parent
     * and a childless one both count 0, and only requiring the hop keeps them apart.
     */
    private fun elementCount(collection: Resolution.Collection, comparison: Threshold): Op<Boolean> {
        val chained = collection.leadingHops.isNotEmpty()
        comparison.vacuous?.let { answer ->
            // A count is never fractional, so the comparison is statically decided — but not
            // unconditionally: a parentless row is a CEL missing-path error, and a constant is
            // two-valued, so `NOT(constant)` would readmit every one of them
            // (cerbos/query-plan-adapters#333). Carrying the guard on an EXPRESSION is what makes
            // both polarities inherit it.
            if (!chained) return if (answer) Op.TRUE else Op.FALSE
            val guarded = translation.subqueries.requireLeadingHops(collection, intLiteral(1), IntegerColumnType())
            return EqOp(guarded, intLiteral(if (answer) 1 else 0))
        }
        // The EXISTS emptiness shortcuts are TWO-valued, so a chain must not take them: `NOT EXISTS`
        // is TRUE for an absent to-one parent, which is why `!(size(chain) > 0)` readmitted every
        // parentless row even though `size(chain) == 0`, guarded separately, did not (#316).
        if (!chained) {
            val nonEmpty = (comparison.operator == "gt" && comparison.value == 0L) ||
                (comparison.operator == "ge" && comparison.value == 1L)
            val empty = (comparison.operator == "eq" && comparison.value == 0L) ||
                (comparison.operator == "le" && comparison.value == 0L) ||
                (comparison.operator == "lt" && comparison.value == 1L)
            if (nonEmpty || empty) {
                val present = translation.subqueries.existsOver(
                    translation.subqueries.chain(collection.owner, collection.hops),
                )
                return if (nonEmpty) present else TriLogic.not(present)
            }
        }
        return compare(translation.subqueries.chainCount(collection), comparison)
    }

    /**
     * `size(collection.filter(x, body))`. CEL's `filter` has NO error absorption: one element whose
     * body errors errors the whole expression, even when the count comparison would otherwise hold.
     * The strict counter goes SQL NULL whenever any element body is UNKNOWN, so every comparison
     * built on it is UNKNOWN and the row stays excluded under both polarities.
     */
    private fun filtered(target: SizeTarget.Filtered, comparison: Threshold, scope: Scope): Op<Boolean> {
        val resolved = scope.resolve(target.variable)
        if (resolved !is Resolution.Collection) {
            throw RelationRefusals.sizeFilterNeedsRelation(target.variable)
        }
        return translation.walker.enterMacro("size(filter(...))") {
            if (comparison.vacuous != null) {
                // The count comparison is statically decided, but an erroring body still has to
                // deny: the poison term alone is 0 when every element is determined and SQL NULL
                // otherwise, which makes the collapse UNKNOWN exactly when CEL errors.
                val poison = translation.subqueries.chainAggregate(resolved, IntegerColumnType()) { alias ->
                    undeterminedPoison(body(resolved, target, scope, alias))
                }
                if (comparison.vacuous) EqOp(poison, intLiteral(0)) else NeqOp(poison, intLiteral(0))
            } else {
                val matched = translation.subqueries.chainAggregate(resolved, IntegerColumnType()) { alias ->
                    val element = body(resolved, target, scope, alias)
                    PlusOp<Int, Int>(
                        Coalesce<Int, Int?>(Sum<Int>(ScoreCase(element, 1, 0, 0), IntegerColumnType()), intLiteral(0)),
                        undeterminedPoison(element),
                        IntegerColumnType(),
                    )
                }
                compare(matched, comparison)
            }
        }
    }

    /**
     * `size([literal list].filter(x, body))`: one body per element POSITION, substituted and walked
     * like the `exists`/`all` fold. The count is `Σ CASE WHEN b THEN 1 ELSE 0 END` plus a poison
     * term per element that is SQL NULL when its body is UNKNOWN — `filter` has no error absorption,
     * so one erroring element errors the size, and NULL is absorbing under `+`.
     */
    private fun filteredLiteral(target: SizeTarget.FilteredLiteral, comparison: Threshold, scope: Scope): Op<Boolean> =
        translation.walker.enterMacro("size(filter(...))") {
            val bodies = target.elements.map { element ->
                translation.walker.traverse(
                    translation.collections.substitute(target.lambda.body, target.lambda.variable, element),
                    scope,
                )
            }
            if (bodies.isEmpty()) {
                // size([]) is 0: decided, whatever the comparison.
                return@enterMacro compare(intLiteral(0), comparison)
            }
            val poison: Expression<Int> = bodies
                .map { nullIfUndetermined(ScoreCase(it, 0, 0, 1)) as Expression<Int> }
                .reduce { a, b -> PlusOp<Int, Int>(a, b, IntegerColumnType()) }
            if (comparison.vacuous != null) {
                return@enterMacro if (comparison.vacuous) EqOp(poison, intLiteral(0)) else NeqOp(poison, intLiteral(0))
            }
            val matched = bodies
                .map { ScoreCase(it, 1, 0, 0) as Expression<Int> }
                .reduce { a, b -> PlusOp<Int, Int>(a, b, IntegerColumnType()) }
            compare(PlusOp<Int, Int>(matched, poison, IntegerColumnType()), comparison)
        }

    private fun body(
        collection: Resolution.Collection,
        target: SizeTarget.Filtered,
        scope: Scope,
        alias: Alias<Table>,
    ): Op<Boolean> = translation.walker.traverse(
        target.lambda.body,
        LambdaScope(translation, alias, collection.tail, target.lambda.variable, scope),
    )

    /** 0 when every element body is determined (and for the empty collection), SQL NULL otherwise. */
    private fun undeterminedPoison(element: Op<Boolean>) = nullIfUndetermined(
        Coalesce<Int, Int?>(Max<Int, Int>(ScoreCase(element, 0, 0, 1), IntegerColumnType()), intLiteral(0)),
    )

    /** A count or a length against the plan's own constant, bound by the VALUE's type. */
    private fun compare(size: Expression<*>, comparison: Threshold): Op<Boolean> {
        val constant = Params.of(comparison.value)
        return when (comparison.operator) {
            "eq" -> EqOp(size, constant)
            "ne" -> NeqOp(size, constant)
            "lt" -> LessOp(size, constant)
            "le" -> LessEqOp(size, constant)
            "gt" -> GreaterOp(size, constant)
            "ge" -> GreaterEqOp(size, constant)
            else -> throw Refusals.malformed("Unsupported size comparison operator: ${comparison.operator}")
        }
    }

    /**
     * What `size()` is applied to. `except()` is called out rather than falling into the generic
     * refusal because it is the PDP-verified wire shape of every real `except()` policy and the
     * rewrite that replaces it is worth naming.
     */
    private sealed interface SizeTarget {
        class Whole(val variable: String) : SizeTarget
        class Filtered(val variable: String, val lambda: ParsedLambda) : SizeTarget
        class FilteredLiteral(val elements: List<Value>, val lambda: ParsedLambda) : SizeTarget

        companion object {
            fun of(argument: Operand): SizeTarget {
                if (argument.nodeCase == Operand.NodeCase.VARIABLE) {
                    return Whole(argument.variable)
                }
                if (argument.nodeCase != Operand.NodeCase.EXPRESSION) {
                    throw RelationRefusals.sizeOperand()
                }
                if (argument.expression.operator == "except") {
                    throw RelationRefusals.exceptUnsupported()
                }
                if (argument.expression.operator != "filter") {
                    throw RelationRefusals.sizeOperand()
                }
                val operands = argument.expression.operandsList
                if (operands.size != 2) {
                    throw Refusals.malformed("filter requires exactly 2 operands, got ${operands.size}")
                }
                if (operands[0].nodeCase != Operand.NodeCase.VARIABLE) {
                    // filter() over a literal list folds per element; over a computed collection
                    // it is legal CEL with no table to count over.
                    val elements = PlanValues.literalListElements(operands[0])
                        ?: throw RelationRefusals.computedCollection("filter")
                    return FilteredLiteral(
                        elements,
                        ParsedLambda.parse(
                            operands[1],
                            "filter second operand must be a lambda",
                            "filter supports single-variable lambdas only",
                            "filter lambda variable must be a variable operand",
                        ),
                    )
                }
                return Filtered(
                    operands[0].variable,
                    ParsedLambda.parse(
                        operands[1],
                        "filter second operand must be a lambda",
                        "filter supports single-variable lambdas only",
                        "filter lambda variable must be a variable operand",
                    ),
                )
            }
        }
    }

    /**
     * A `size()` comparison reduced to an integral threshold.
     *
     * A count and a character length are both integral, so a fractional constant never lands on
     * one: `eq 1.5` is always false, `ne 1.5` always true, and `>= 1.5` is `>= 2` — rounding, not
     * truncating. Truncating `>= 1.5` to `>= 1` returns rows the PDP denies, which is the bug the
     * corpus's fractional actions exist to catch.
     */
    private class Threshold(val operator: String, val value: Long, val vacuous: Boolean?) {
        companion object {
            fun of(operator: String, raw: Double): Threshold {
                // BEFORE the rint test, which NaN fails — `NaN == rint(NaN)` is false — so it fell
                // into the fractional arms, where `ceil(NaN).toLong()` is 0 and `size(x) > NaN`
                // became the always-true `size(x) >= 0`.
                if (!raw.isFinite()) {
                    if (raw.isNaN()) throw RelationRefusals.nanSizeThreshold()
                    // An INFINITY is decided, and decided correctly, by the out-of-range arms
                    // below: IEEE orders the infinities totally, so `size(x) > +Infinity` is false
                    // and `size(x) < +Infinity` is true for every present value, which is exactly
                    // what a bound past `Int.MAX_VALUE` already answers. The saturation `toLong`
                    // performs is the intended step here rather than an accident of narrowing.
                    return Threshold(operator, raw.toLong(), null)
                }
                if (raw == Math.rint(raw)) {
                    return Threshold(operator, raw.toLong(), null)
                }
                return when (operator) {
                    "eq" -> Threshold(operator, 0, false)
                    "ne" -> Threshold(operator, 0, true)
                    // An integral size makes gt and ge coincide once the bound is rounded, as do
                    // lt and le.
                    "gt", "ge" -> Threshold("ge", ceil(raw).toLong(), null)
                    "lt", "le" -> Threshold("le", floor(raw).toLong(), null)
                    // size() yields an int, so anything but a comparison over it is a CEL type
                    // error the planner would not have shipped.
                    else -> throw Refusals.malformed("Unsupported size comparison operator: $operator")
                }
            }
        }
    }
}
